reframe/core/containers.py

```diff
@@ -88,25 +110,28 @@ def validate(self):
         if self.image is None:
             raise ContainerError('no image specified')
 
-        if not self.commands:
-            raise ContainerError('no commands specified')
-
 
 class Docker(ContainerPlatform):
     '''Container platform backend for running containers with `Docker
     <https://www.docker.com/>`__.'''
 
     def emit_prepare_commands(self):
+        if self.pull_command == '':
+            return [f'docker pull {self.image}']
+        elif self.pull_command:
+            return [self.pull_command]
+
         return []
 
     def launch_command(self):
         super().launch_command()
-        run_opts = ['-v "%s":"%s"' % mp for mp in self.mount_points]
+        run_opts = [f'-v "{mp[0]}":"{mp[1]}"' for mp in self.mount_points]
+        run_opts += ['-v "${PWD}":"/rfm_workdir"']
         run_opts += self.options
-        run_cmd = 'docker run --rm %s %s bash -c ' % (' '.join(run_opts),
-                                                      self.image)
-        return run_cmd + "'" + '; '.join(
-            ['cd ' + self.workdir] + self.commands) + "'"
+        workdir_opt = f'--workdir="{self.workdir}" ' if self.workdir else ''
+
+        return (f'docker run --rm {workdir_opt}{" ".join(run_opts)} '
```

Comment:
Thinking about it; in the future you might want to change that ti use the docker-py SDK to create containers within python. Even though that would break with all the other execution environments.